decentraland/explorer
# feat: enable stateful workers

## What are we doing?
Our current SDK has no concept of state. It simply generates messages that are sent to the renderer, so that it updates the scene in some way. This works, but it has some disadvantages. As part of the builder in-world feature, we want to start considering the scene's state as a first class citizen, and share it across players. By doing so, we will be able to support multiplayer world edit tools, such as the builder in-world.

### General idea
Right now, each scene runs in a different webworker. We decided to create a new webworker that runs completely independent from the others, so that we can switch between them when the user enters and leaves edit mode. The new webworker will contain the scene's state, and keep it in sync with the renderer. It will also be in charge of saving and retrieving the state on a semi-persistent storage.

### Implementation details
We took the current `SceneWorker` and created an abstraction for it. This implied hiding a lot of exposed properties behind class methods. Then, we took the current implementation and called it `SceneSystemWorker`. The new worker is called `StatefulWorker`.

When the user enters edit mode, it will send a message to kernel. Kernel will then kill the scene worker, and start a new stateful one. The opposite will happen when the users leave edit mode. 

We are now storing the state in local storage. This was mainly done to enable an easy integration with the builder in-world feature, without needing to have deploy access to a LAND. The state is stored in a json format, in an easy to read format. We could change this in the future if we want to, but until the feature is stable enough, it will make things easier.

## How to test it locally
1. Run kernel 
```make watch```
2. Go to http://localhost:8080/?DEBUG_MODE&position=-1%2C-51. You will see a regular scene being executed.
3. Open the console and execute
```var sceneId = 'Qm2uuMLmJKSZ1Z4PUnDgxqnvWnVTMGDaJmdke98cTKWHSxVgg649rTiMZojqcTj2ft1X4mb7uZSn11Xi7LBBMtUjTV'```
4. Finally enter stateful mode by executing on the console:
```browserInterface.ControlEvent({eventType: 'StartStateMode', payload: {sceneId}})```
You will see how the scene unloaded, and a new version of the scene loads. That version is defined in a descriptive way.
_Note: the event type was changed to 'StartStatefulMode' in https://github.com/decentraland/explorer/pull/1589_ 

From there on, you can trigger the following actions:
* Leave edit mode:
```browserInterface.ControlEvent({eventType: 'StopStateMode', payload: {sceneId}})```
_Note: the event type was changed to 'StopStatefulMode' in https://github.com/decentraland/explorer/pull/1589_ 
* Add entities:
```browserInterface.SceneEvent({sceneId, eventType:'stateEvent', payload: { type: 'AddEntity', payload: { entityId: 'E2', components: [{ id: 1, data: { position: { x: 8, y: 0, z: 15 }}}, { id: 54, data:{ src: 'models/BlockDog.glb' } }]}}})```
* Modify components:
```browserInterface.SceneEvent({sceneId, eventType:'stateEvent', payload: { type: 'SetComponent', payload: { entityId: 'E1', componentId: 1, componentData: {position: { x: 8, y: 0, z: 1 }}}}})```
* Remove entities:
```browserInterface.SceneEvent({sceneId, eventType:'stateEvent', payload: { type: 'RemoveEntity', payload: { entityId: 'E2' }}})```
* Remove components:
```browserInterface.SceneEvent({sceneId, eventType:'stateEvent', payload: { type: 'RemoveComponent', payload: { entityId: 'E1', componentId: 1 }}})```
* Store the current state in local storage:
```browserInterface.SceneEvent({sceneId, eventType:'stateEvent', payload: { type: 'StoreSceneState', payload: { }}})```

It is important to mention that whenever you execute one of the above, you won't see it reflected on the renderer. This is because we are simulating changes that happen on the renderer, so it makes no sense to send them back. If you want to see the changes reflected, then you have to:
1. Save the changes
2. Reload the scene and enter edit mode, or directly enter edit mode again

## Open questions
**Components support**
With this PR, we are only supporting two types of components: `Transform` and `GLTFShape`. The current builder also supports `Script` and `NFTShape`. Which ones do we want to support for the builder in-world feature?

**Files**
We need to figure out how to handle files, such as gtlfs. We are currently only syncing state data between kernel/worker & the renderer

**Component data**
We need to agree on what the renderer will send to the worker. Will it send the whole component, or will it send just the updates? We should use the same logic for syncing between players in the future.

## Next steps
* Agree on open questions and handle those changes in future PRs
* Start sending the scene's state to content server (when not in `DEBUG_MODE`)
* Generate the scene's code, so that it is compatible with the current way of executing a scene
